test/test_functest.py

```diff
@@ -5,87 +5,87 @@
 
 import os
 import platform
+import pytest
 import unittest
 
 import pqclean
 import helpers
 
 
-def test_functest():
-    for scheme in pqclean.Scheme.all_schemes():
-        for implementation in scheme.implementations:
-            yield check_functest, implementation
-
-
-def test_functest_sanitizers():
-    for scheme in pqclean.Scheme.all_schemes():
-        for implementation in scheme.implementations:
-            yield check_functest_sanitizers, implementation
-
-
+@pytest.mark.parametrize(
+    'implementation,test_dir,impl_path, initializer, destructor',
+    [(impl, *helpers.isolate_test_files(impl.path(), 'test_functest_'))
+     for impl in pqclean.Scheme.all_implementations()],
+    ids=[str(impl) for impl in pqclean.Scheme.all_implementations()],
+)
 @helpers.filtered_test
-def check_functest(implementation):
+def test_functest(implementation, impl_path, test_dir,
+                  initializer, destructor):
+    initializer()
+    dest_dir = os.path.join(test_dir, 'bin')
     helpers.make('functest',
                  TYPE=implementation.scheme.type,
                  SCHEME=implementation.scheme.name,
                  IMPLEMENTATION=implementation.name,
-                 working_dir=os.path.join('..', 'test'))
+                 SCHEME_DIR=impl_path,
+                 DEST_DIR=dest_dir,
+                 working_dir=os.path.join(test_dir, 'test'))
     helpers.run_subprocess(
-        [os.path.join('..', 'bin', 'functest_{}_{}{}'.format(
+        [os.path.join(dest_dir, 'functest_{}_{}{}'.format(
             implementation.scheme.name,
             implementation.name,
             '.exe' if os.name == 'nt' else ''
         ))],
-        os.path.join('..', 'bin'),
     )
+    destructor()
 
 
-@helpers.filtered_test
+@pytest.mark.parametrize(
+    'implementation,test_dir,impl_path, initializer, destructor',
+    [(impl,
+      *helpers.isolate_test_files(impl.path(), 'test_functest_sanitizers_'))
+     for impl in pqclean.Scheme.all_implementations()],
+    ids=[str(impl) for impl in pqclean.Scheme.all_implementations()],
+)
 @helpers.skip_windows()
+@helpers.filtered_test
 @helpers.slow_test
-def check_functest_sanitizers(implementation):
+def test_functest_sanitizers(implementation, impl_path, test_dir,
+                             initializer, destructor):
+    dest_dir = os.path.join(test_dir, 'bin')
     env = None
     if platform.machine() == 'ppc' and os.environ.get('CC', 'gcc') == 'clang':
         raise unittest.SkipTest("Clang does not support ASAN on ppc")
     elif platform.machine() in ['armv7l', 'aarch64']:
         env = {'ASAN_OPTIONS': 'detect_leaks=0'}
     elif platform.system() == 'Darwin':
-        raise unittest.SkipTest('valgrind is not reliable on OSX')
+        raise unittest.SkipTest('ASAN is not reliable on OSX')
     else:
         print("Supported platform: {}".format(platform.machine()))
 
+    initializer()
```

Comment:
rename `init()` as in other tests?
Also should this be called in all tests or is it only required for some?